cpp/drogon/CMakeLists.txt

```diff
@@ -0,0 +1,110 @@
+cmake_minimum_required (VERSION 3.2)
+Project(drogon_benchmark)
+
+link_directories(/usr/local/lib)
+link_libraries(drogon trantor pthread dl)
+
+IF (CMAKE_SYSTEM_NAME MATCHES "Linux")
```

Comment:
It's very bad form to put toolchain related code in the cmakelists.txt file, this kind of stuff belongs in the toolchain.  The cmakelists.txt file should not contain anything like this, the most it should contain is the feature set request, and if you want to mandate the C++ version then you should explicitly state so, not pass compiler flags, so something [like this](https://cmake.org/cmake/help/v3.15/prop_tgt/CXX_STANDARD.html#prop_tgt:CXX_STANDARD):
```cmakelist
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
`CMAKE_CXX_STANDARD` will attempt to request that version (and there are detailed feature tests if you need those as well, but you always don't), and `CMAKE_CXX_STANDARD_REQUIRED` means that the standard version specified is hard required and that the code will not fall back on older versions.

Reply:
@OvermindDL1 , thank you for your explanation, but I still have a question, my framework needs to use the `any`, I need to choose std::any or std::experimental::any according to the standards supported by the compiler, so I need to detect the version of the compiler. How can I do the above things while avoiding the wrong practices you pointed out? I don't know much about this, thank you for your correction.

Reply:
@an-tao Well first of all the Standards namespace should always be used, `std::any`, any compiler that doesn't support that isn't going to support much of anything at this point (even MSVC has full C++17 support, GCC and clang both long before that), so just setting the standards version and using `std::any` is always preferred.  Second, the `std::experimental` namespace should never, absolutely Never be used in a production product.  And third, if you really must choose between them then you 'could' use a compiler (like gcc/clang) specific way to test (a macro), but that's not portable, or the proper way with CMake would be to setup a Feature test, which can either be a built-in test (don't think `any` is standalone as it is covered by C++17, so you really really should just use that), or built your own (where you just make a minimal cpp file that fails to compile if the feature doesn't exist, I.E. just try including the proper header, then cmake can set a DEFINE on if that succeeds or not, which your code can conditionally compile differently).

Again though, I would strongly *strongly* recommend never Ever to use anything from `std::experimental` in a distributed or production set of code.  Just mandate C++17 minimum, every modern compiler supports it and has for years now.

If you really want an `any` type but want it to work on pre-C++17 versions, then I'd recommend using `boost::any` (which you can include via pure cmake with a [hunter.sh](https://hunter.sh) cmake file if you want to be particularly simple, or just strip it out of boost as boost has specific tools to do just that), but still don't use `std::experimental`.

Reply:
@OvermindDL1 Thank you very much. Your answer is very clear and reasonable. I will remove std::experimental as soon as possible. In fact, some old projects (like [this](https://github.com/an-tao/drogon/issues/157)) makes me have to consider supporting compilation with C++14.
I think boost::any is a good alternative. Would you like to review the code for me after I made the modification? Thank you very much!

Reply:
I'd be up for that, just ping me on whatever you need.  I've been missing C++ (been mostly in OCaml/Rust/Elixir lately because of my day job but C++ has been my life-long love).  :-)

Reply:
@OvermindDL1 thank you very much! Today I tried to strip out any and string_view from boost by it's tool `bcp`, but more than 700 files were copied to my target folder, the files number is too large for me. Maybe the method of using bcp is wrong, I will study it tomorrow.
I have never used hunter.sh, I have to take some time to study it.
As for trantor, I have done some benchmarks against libevent, However, the tests I did were not comprehensive, and from the results, the performance of trantor is slightly better. I believe that the API with modern C++ is certainly more friendly to the user than C, so I think trantor is easier to use than libevent. However, as far as the API is concerned, the asio is obviously better because it is compatible with the C++ networking TS  and is likely to be received by the new standard.

Reply:
> Today I tried to strip out any and string_view from boost by it's tool bcp, but more than 700 files were copied to my target folder, the files number is too large for me.

Knowing boost that might not actually be wrong...  ^.^;

However, I bet it can still be slimmed down a lot.  `bcp` has a number of options for that.

> I have never used hunter.sh, I have to take some time to study it.

It's basically a simplified dependency management system completely built in cmake (with shared cache, etc... etc... the usual things you'd expect, made by one of the big known CMake people).

In essence copy the 'huntergate' module into your project, traditionally into the `cmake` folder (the usual place to put cmake modules, but that's only convention and not a hard requirement), a simple command if you are in your root project folder is just `mkdir -p cmake; wget https://raw.githubusercontent.com/hunter-packages/gate/master/cmake/HunterGate.cmake -O cmake/HunterGate.cmake`, then to use it just add this to the base-most cmake file (the `CMakeLists.txt` file in the root project directory) to load and initialize the module:
```cmake
include("cmake/HunterGate.cmake")
```
Then just have it load a metadata repository version (which stays cached locally), like this is the latest (you don't need the latest, just whichever that contains the metadata of what you want, you *don't* need to keep it updated unless you want something newly added, this is immutable and static for a reason, build reproducibilty):
```cmake
HunterGate(
    URL "https://github.com/ruslo/hunter/archive/v0.23.204.tar.gz"
    SHA1 "32cfed254da901f6f184027d530d8da47e035b85"
)
```
And of course the user of the project can specify arguments to their `cmake` command to disable it, have it use the system instead, etc... etc..., but by default it will acquire and build whatever you request, which for boost::any (which is included in the default boost header set with no compile-time components needed):
```cmake
hunter_add_package(Boost)
```
You can also say to use a specific version of boost, otherwise the above will use the latest version of boost in the given metadata repository file, I generally prefer to put a version, so like this instead:
```cmake
hunter_config(
    Boost
    VERSION ${HUNTER_Boost_VERSION}
)
```
Where earlier you can set the `${HUNTER_Boost_VERSION}` variable to whatever boost version as a default value (not a hard set value, I like to use variables for it as it lets the user of your library override it with ease, although with boost::any being such a basic unchanging thing just leaving it versionless may be fine, but still, I opt for versions).

To use it you use it like any other library (hunter manages adding the proper paths for the dependency it acquired and everything, so all of this "just works" without you needing to supply any Find*.cmake files or anything):
```cmake
# Find it and require it to exist, using `CONFIG` here to get automatic source
# mapping and all, this is normal CMake stuff, nothing special.
find_package(Boost CONFIG REQUIRED)

# Then link the headers into your project as normal, just add `Boost::boost` to
# your target_link_libraries line.
target_link_libraries(... Boost::boost)
```

And of course whatever other libraries hunter.sh's default metadata repository has then you can depend on them that way as well, or you can go through the work (bit of trouble as it has a lot of steps to confirm everything works right with tests and all) to put your libraries in hunter.sh if you want as well (all documented in the docs, definitely not easy like rust's cargo/crate manager, but very reliable).

Different patterns, different choices, and of course there is a lot more as well.  ^.^

Reply:
Thank you so much, your introduction is very clear.

According to my understanding, hunter.sh will download the corresponding library to the build directory instead of installing it into the system, so if `trantor` uses hunter.sh, then the application that uses `trantor` must also use hunter.sh for the libraries needed by trantor, otherwise errors may occur because libraries were not found or different version of libraries in system were used. Users of trantor must know how to use hurter.sh correctly, which is not friendly to the user.
So is hunter.sh more suitable for building applications than libraries? This makes me a bit confused. If I understand it incorrectly, please correct me. Thank you

Reply:
> hunter.sh will download the corresponding library to the build directory instead of installing it into the system

Actually by default it downloads it to something like `$HOME/.hunter` or so to be used as a global cache directory to save repeated compilation time (although there is an parameter to configure that as I recall).

> so if trantor uses hunter.sh, then the application that uses trantor must also use hunter.sh for the libraries needed by trantor, 

Nope, hunter is not 'infective', you can use as little or as much as you want, further dependencies do not need to use it, rather only 'your' library in their project will use it, and they can of course disable it to manually manage the library installation as well.

> Users of trantor must know how to use hurter.sh correctly, which is not friendly to the user.

Not at all, in 99% of cases users of your library won't know that hunter even exists.  It is designed to be as transparent as possible for users.

> So is hunter.sh more suitable for building applications than libraries? This makes me a bit confused. If I understand it incorrectly, please correct me. Thank you

It is suitable for any stage, and if used in libraries it is essentially invisible to them unless they actively pass the flag to disable it in your library.  In essence, using it means you don't need to put a Boost dependency (or whatever else you want hunter to acquire) in your readme, though it might be useful to say that you auto-acquire a fully embedded boost and if you want the end-user to manage installing it manually instead then they can pass in `-DHUNTER_ENABLED=NO` to their cmake invocation and everything about hunter is just gone, meaning they have to acquire boost, they have to setup the paths for it again, with a version that is compatible with your library, etc...  There is a list of user flags at https://docs.hunter.sh/en/latest/reference/user-variables.html if you are curious of the usual set.

Hunter is not like conan or vcpkg, it does not require installing secondary software or running pre-compilation steps, nor does it require the user to even know it exists for the 99% of usages when it is used in a library, it's designed to be as transparent, yet overrideable as possible.

Reply:
@OvermindDL1 Thanks a lot!

Currently, the usage of trantor is as follows:
* git clone ....trantor
* cmake ..
* make && make install(copies the header files and library files to the system)
* and then the user creates his own project, includes trantor header files, compiles and links the library files to generate their applications. 
But trantor has the following interfaces:
https://github.com/an-tao/trantor/blob/b2162c924ad4d2678fdfa813e021620bc1044d94/trantor/net/TcpConnection.h#L62-L66
As you can see, these interfaces contain references to std::any or boost::any, if the boost is download by hunter,  It must also be installed into the trantor installation directory (maybe there is another version of Boost in system, so trantor's boost should be installed into trantor's folder), otherwise the user must pay attention to the boost dependency problem. Do I understand this right?

Thank you very much for your introduction, I have learned a lot.

Reply:
> Currently, the usage of trantor is as follows:
>
> * git clone ....trantor
> * cmake ..
> * make && make install(copies the header files and library files to the system)
> * and then the user creates his own project, includes trantor header files, compiles and links the library files to generate their applications.

This wouldn't change.

> As you can see, these interfaces contain references to std::any or boost::any, if the boost is download by hunter, It must also be installed into the trantor installation directory (maybe there is another version of Boost in system, so trantor's boost should be installed into trantor's folder), otherwise the user must pay attention to the boost dependency problem. Do I understand this right?

If you expose boost then you either need to have the user depend on boost (whether system or via hunter as well, you can make a simple cmake module they can include to auto-include whatever necessary if you want to make it easy), or you need to vendor in the boost parts (bcp again).

For public interfaces I would opt for bcp to include it internally.  But then again I highly question using `any` here at all, instead the TCPConnection class should be templated on the context type instead.  That's a decent bit of processing overhead for no reason.

Reply:
Yes, the template type sounds better, how could I miss this! 

Thank you very much. ^_^

Reply:
@OvermindDL1, I have made a [PR](https://github.com/an-tao/trantor/pull/40) on trantor to remove `any`, but I don't know how to ask you to review it, when I am going to do this in the right sidebar, there is no button to invite you.

<img width="359" alt="屏幕快照 2019-07-20 下午12 29 58" src="https://user-images.githubusercontent.com/21002501/61574055-8829fb80-aaeb-11e9-9d5e-7394ab2a461b.png">

Regarding this PR, I sent an email to your gmail email. I would be grateful if I could get your suggestion, thank you.

Reply:
@an-tao To add @OvermindDL1 as reviewer, you need to add him as collaborator first ^^
https://help.github.com/en/articles/inviting-collaborators-to-a-personal-repository :heart: 

BTW, very participative and clear discussion here :+1:

Reply:
@waghanza Thanks a lot.

Reply:
> @OvermindDL1, I have made a PR on trantor to remove any, but I don't know how to ask you to review it, when I am going to do this in the right sidebar, there is no button to invite you.

Or just ping me in the comments and I can look at it, I'm pretty easy.  ^.^